tfhe/src/typed_api/integers/client_key.rs

```diff
@@ -1,55 +1,147 @@
+            Self {
+                inner,
+                pbs_order: params.pbs_order,
+            }
+        }
+        #[cfg(not(feature = "internal-keycache"))]
+        {
+            let inner =
+                crate::integer::RadixClientKey::new(params.block_parameters, params.num_block);
+            Self {
+                inner,
+                pbs_order: params.pbs_order,
+            }
+        }
+    }
+}
 
-    fn encrypt(&self, value: U256) -> Self::Ciphertext {
-        self.as_ref().encrypt_radix(value, self.num_blocks())
+impl EncryptionKey<u64, RadixCiphertextBig> for RadixClientKey {
+    fn encrypt(&self, value: u64) -> RadixCiphertextBig {
+        assert_eq!(
+            self.pbs_order,
+            crate::shortint::PBSOrder::KeyswitchBootstrap
+        );
+        self.inner.encrypt(value)
     }
 }
 
-impl DecryptionKey<u64> for RadixClientKey {
-    type Ciphertext = RadixCiphertextBig;
+impl EncryptionKey<U256, RadixCiphertextBig> for RadixClientKey {
+    fn encrypt(&self, value: U256) -> RadixCiphertextBig {
+        self.inner
+            .as_ref()
+            .encrypt_radix(value, self.inner.num_blocks())
+    }
+}
 
-    fn decrypt(&self, ciphertext: &Self::Ciphertext) -> u64 {
-        self.decrypt(ciphertext)
+impl EncryptionKey<u64, RadixCiphertextDyn> for RadixClientKey {
+    fn encrypt(&self, value: u64) -> RadixCiphertextDyn {
+        match self.pbs_order {
+            crate::shortint::PBSOrder::KeyswitchBootstrap => {
+                RadixCiphertextDyn::Big(self.inner.encrypt(value))
+            }
+            crate::shortint::PBSOrder::BootstrapKeyswitch => RadixCiphertextDyn::Small(
+                self.inner
+                    .as_ref()
+                    .encrypt_radix_small(value, self.inner.num_blocks()),
+            ),
+        }
     }
 }
 
-impl DecryptionKey<U256> for RadixClientKey {
-    type Ciphertext = RadixCiphertextBig;
+impl EncryptionKey<U256, RadixCiphertextDyn> for RadixClientKey {
+    fn encrypt(&self, value: U256) -> RadixCiphertextDyn {
+        match self.pbs_order {
+            crate::shortint::PBSOrder::KeyswitchBootstrap => RadixCiphertextDyn::Big(
+                self.inner
+                    .as_ref()
+                    .encrypt_radix(value, self.inner.num_blocks()),
+            ),
+            crate::shortint::PBSOrder::BootstrapKeyswitch => RadixCiphertextDyn::Small(
+                self.inner
+                    .as_ref()
+                    .encrypt_radix_small(value, self.inner.num_blocks()),
+            ),
+        }
+    }
+}
 
-    fn decrypt(&self, ciphertext: &Self::Ciphertext) -> U256 {
+impl DecryptionKey<RadixCiphertextBig, u64> for RadixClientKey {
+    fn decrypt(&self, ciphertext: &RadixCiphertextBig) -> u64 {
+        assert_eq!(
+            self.pbs_order,
+            crate::shortint::PBSOrder::KeyswitchBootstrap
+        );
+        self.inner.decrypt(ciphertext)
+    }
+}
```

Comment:
you can't match on the variant like above ? Or maybe I'm misreading something

Reply:
We can't  here, we use a radix key to encrypt a big radix, the assert is there to check that it matches the fact
that the client key was created with small params

Reply:
then my question is rather : why do we have a specialized implementation for the Big variant ?

Reply:
it was needed to be able to incrementally makes things works, it is not needed now so can be removed

Reply:
Ok, I think a ctrl+f for those "specialized" impls can be worth a go